pageserver/src/tenant/timeline.rs

```diff
@@ -575,6 +642,54 @@ impl Timeline {
         res
     }
 
+    pub(crate) const MAX_GET_VECTORED_KEYS: u64 = 32;
+
+    /// Look up multiple page versions at a given LSN
+    ///
+    /// This naive implementation will be replaced with a more efficient one
+    /// which actually vectorizes the read path.
+    pub(crate) async fn get_vectored(
+        &self,
+        key_ranges: &[Range<Key>],
+        lsn: Lsn,
+        ctx: &RequestContext,
+    ) -> Result<BTreeMap<Key, Result<Bytes, PageReconstructError>>, GetVectoredError> {
+        if !lsn.is_valid() {
+            return Err(GetVectoredError::InvalidLsn(lsn));
+        }
+
+        let key_count = key_ranges
+            .iter()
+            .map(|range| key_range_size(range) as u64)
+            .sum();
+        if key_count > Timeline::MAX_GET_VECTORED_KEYS {
+            return Err(GetVectoredError::Oversized(key_count));
+        }
+
+        let mut values = BTreeMap::new();
+        for range in key_ranges {
+            let mut key = range.start;
+            while key != range.end {
+                debug_assert!(!self.shard_identity.is_key_disposable(&key));
+
+                let block = self.get(key, lsn, ctx).await;
+
+                if matches!(
+                    block,
+                    Err(PageReconstructError::Cancelled)
+                        | Err(PageReconstructError::AncestorStopping(_))
```

Comment:
Manual clippy notice, this can now be done as the next hunk already has:

```suggestion
                    Err(PageReconstructError::Cancelled | PageReconstructError::AncestorStopping(_))
```

In some cases it might yield nicer formatting, but probably not here.

Reply:
Done in [add87fc](https://github.com/neondatabase/neon/pull/6372/commits/add87fc6a288fce6b91233f28b5d4b3a87b18af8)